v3/src/components/graph/components/graph.tsx

```diff
@@ -188,6 +189,7 @@ export const Graph = observer(function Graph({graphController, graphRef, dotsRef
           />
         </svg>
         {renderDroppableAddAttributes()}
+        {graphModel.adornments.length > 0 && <Adornments />}
```

Comment:
```suggestion
        <Adornments />
```
The `Adornments` component should be responsible for deciding what to render when there are no adornments.